daemon/api.go

```diff
@@ -2806,3 +2852,81 @@ func systemCoreSnapUnalias(name string) string {
 	}
 	return name
 }
+
+func listSnapshots(c *Command, r *http.Request, user *auth.UserState) Response {
+	query := r.URL.Query()
+	var id uint64
+	sid := query.Get("id")
+	if sid != "" {
+		var err error
+		id, err = strconv.ParseUint(sid, 10, 64)
+		if err != nil {
+			return BadRequest("'id', if given, must be a positive base 10 number; got %q", sid)
+		}
+	}
+
+	shots, err := snapshotList(context.TODO(), id, splitQS(r.URL.Query().Get("snaps")))
+	if err != nil {
+		return InternalError(err.Error())
+	}
+	return SyncResponse(shots, nil)
+}
+
+func changeSnapshots(c *Command, r *http.Request, user *auth.UserState) Response {
+	var action client.SnapshotAction
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&action); err != nil {
+		return BadRequest("cannot decode request body into snapshot operation: %v", err)
+	}
+	if decoder.More() {
+		return BadRequest("extra content found after snapshot operation")
+	}
+
+	if action.ID == 0 {
+		return BadRequest("snapshot operation requires snapshot ID")
+	}
+
+	if action.Action == "" {
+		return BadRequest("snapshot operation requires action")
+	}
+
+	var affected []string
+	var ts *state.TaskSet
+	var err error
+
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	switch action.Action {
+	case "check":
+		affected, ts, err = snapshotCheck(st, action.ID, action.Snaps, action.Users)
+	case "restore":
+		affected, ts, err = snapshotRestore(st, action.ID, action.Snaps, action.Users)
+	case "forget":
+		if len(action.Users) != 0 {
+			return BadRequest(`snapshot "forget" operation cannot specify users`)
+		}
+		affected, ts, err = snapshotForget(st, action.ID, action.Snaps)
+	default:
+		return BadRequest("unknown snapshot operation %q", action.Action)
+	}
+
+	switch err {
+	case nil:
+		// woo
+	case client.ErrSnapshotNotFound, client.ErrSnapshotSnapsNotFound:
+		return NotFound(err.Error())
+	default:
+		return InternalError("%v", err)
```

Comment:
not right because of consistency? Probably right; we've got 12 of the `err.Error()` variants and 16 of the `"%v", err` variants.
Not right in the practical sense? not really: errors are `func(string, ...interface{})`, and `%v` of a non-nil `error` calls `Error()` on it.

Reply:
Not right in practice.. the first parameter seems to be a formatting string, so it's a real bug to feed an undetermined string into it. In C that can easily lead to a crash. In Go we'll just get a corrupted string. Given we have 12 cases incorrect, perhaps we should drop that formatting logic and just let the call site use Sprintf when necessary. Alternatively, we could make these parameters use the _error_ type, so it's extremely convenient for the typical case, and we can use fmt.Errorf otherwise.

Reply:
d'oh, you are absolutely correct, and that's the underlining error we had (have, still) in the description output of snap info...

Reply:
I'll look at changing it to take an `error`, but in another PR.